src/lib/fsm/index.ts

```diff
@@ -0,0 +1,54 @@
+export default class fsm {
+  private currentState: string;
+  private states: { [id: string]: { [id: string]: string } } = {};
+
+  constructor(initialState: string, transitions: Array<transition>) {
+    this.currentState = initialState;
+
+    transitions.forEach((transition) => this.addTransition(transition));
+  }
+
+  public CurrentState(): string {
+    get: {
+      return this.currentState;
+    }
+  }
+
+  public handle(message: string) {
+    if (this.states[this.currentState] != undefined) {
+      var target = this.states[this.currentState][message];
+
+      if (target != undefined) {
+        this.currentState = target;
+      }
+    }
+  }
+
+  public static Transition(fromState: string, message: string, toState: string): transition {
+    return new transition(fromState, message, toState);
+  }
+
+  private addTransition(transition: transition) {
+    let fromState = transition.fromState;
+    let message = transition.message;
+    let toState = transition.toState;
+
+    if (this.states[fromState] == undefined) {
+      this.states[fromState] = {};
+    }
+
+    this.states[fromState][message] = toState;
```

Comment:
mutable state (its plain js object) use immutable:

this.states[fromState] = { ...this.states[fromState], message: toState };

But I would rather have Types/Classes to define "Machine" + "Node" + "Transition" + "State". (Note that node and state are different).

We are probably also overriding if repeated definition of from+message -> should throw an error